### PR TITLE
fixed initial draw of columns when tree contains hidden nodes

### PR DESCRIPTION
**Pull request by PetzeltA**
setting gridCellPrevId, gridCellNextId and gridCellChildId outside the column loop speeds up the code a bit

### CODE REVIEW

**Comment by deitch** [comment]
So this one resolves #229 ?

**Comment by deitch** [comment]
Definitely we should have kept these as separate PRs. It makes it easier to wrap one's head around.

Let's deal with them separately. First, can you pull the indicated lines into their own branch and make a PR on it? That one we can merge in immediately. Then we can get our heads around this change.

File: jstreegrid.js [hunk 1]
```diff
@@ -1035,6 +1035,31 @@
...
 				// find which column our tree shuld go in
 				var s = this.settings.grid;
+
+				var prevPos = pos - 1;
+
+				while (prevPos >= 0 && this.is_hidden(peers[prevPos])) {
+					prevPos--;
+				}
+
+				gridCellPrevId = prevPos < 0 ? objData.parent : findLastClosedNode(this, peers[prevPos]);
```
Comment (deitch) [inline jstreegrid.js]: This line...

File: jstreegrid.js [hunk 2]
```diff
@@ -1035,6 +1035,31 @@
...
+				var nextPos = pos + 1;
+				gridCellNextId = pos >= peers.length - 1 ? "NULL" : peers[nextPos];
```
Comment (deitch) [inline jstreegrid.js]: ... this line...

File: jstreegrid.js [hunk 3]
```diff
@@ -1035,6 +1035,31 @@
...
+				while (gridCellNextId !== "NULL" && this.is_hidden(gridCellNextId)) {
+					gridCellNextId = nextPos >= peers.length - 1 ? "NULL" : peers[++nextPos];
+				}
+
+				var childPos = 0;
+				gridCellChildId = "NULL";
```
Comment (deitch) [inline jstreegrid.js]: .. and this line (with the appropriate code) should be in its own PR. Let's keep the logic for performance enhancements and logic for bugfixes separate.

**Comment by deitch** [comment]
Then we can keep this PR for the bug fix.

**Comment by deitch** [comment]
OK, I get the performance improvements. Basically take the calcs that will be identical across all columns and extract them. The only thing that needs to be done for _each_ column is finding the actual data cells. Makes sense.

Do a separate simple PR and we will merge it in immediately. :-)

**Comment by deitch** [comment]
On this one (fixing the display when some nodes are hidden), can you update the documentation at https://github.com/PetzeltA/jstree-grid/blob/4bda2521ed58476be77f444c8d1e716658e58624/jstreegrid.js#L1157 to describe how the new logic works? It probably should go higher up where you put the modified logic.

The original logic is:

1. We are the first child: install after the parent
2. Our previous peer is already drawn: install after the previous peer
3. Our previous peer is not drawn, we have a child that is drawn: install right before our first child
4. Our previous peer is not drawn, we have no child that is drawn, our next peer is drawn: install right before our next peer
5. Our previous peer is not drawn, we have no child that is drawn, our next peer is not drawn: install right after parent

It looks like the new logic is the same, but with some changes:

1. When determining the previous peer, look for the previous _visible_ peer (skip hidden)
2. When determining our next peer, look for our next _visible_ peer (skip hidden)
3. When determining if we have children, look only for _visible_ children

If I understood correctly, the net of this is that the position for each node will be set only by visible nodes; invisible ones will not be rendered, or if they are, in strange locations, but since they are visible, it should not matter.

I wonder if the more correct logic might be to account for all nodes, but just to keep the hidden ones hidden? So if we have 4 nodes in a row - A hidden, B visible, C hidden, D visible - then we still render A->B->C->D in that order, bur since A & C are hidden, only B & D show.

Besides being a little easier to understand - everything stays in the same order - when a node switches to visible, it makes it easier to handle: just show it.

My only question then is why it doesn't work this way now....

**Comment by PetzeltA** [comment]
Sorry - I don't know what to do now. The changes in the calculation fixed the loading problem (though there still is a problem when sorting (see API show/hide test and sorting by price).
Now - as far as I understand - You want me to change the position of the calculation do it's original line-positoin within the loop. Then I sould do a second PR which moves the lines to where they are now?

I am currently working at a better solution.
The grid-items should not be removed but hidden - as in the oriignal control. This makes also the new position logic obsolete. It also should solve the sorting problem .

**Comment by deitch** [comment]
> Sorry - I don't know what to do now

Ok, no worries, I don't need to be a purist about the PRs. :-)

**Comment by deitch** [comment]
> The changes in the calculation fixed the loading problem (though there still is a problem when sorting

I will look, but if:

> I am currently working at a better solution. The grid-items should not be removed but hidden

Yeah, that is exactly what I was asking. Let us match the tree nodes.

**Comment by PetzeltA** [comment]
What I'd like to do is to revert all changes to the point where I first found the issues with the hidden nodes.
Then I'd do a clean fork and start over again doing 2 PRs.
1. dealing with the hidden nodes issue by hinding them
2. seeding up drawing by detaching the columns from the dom and reattaching them after the drawing operation

**Comment by deitch** [comment]
> Then I'd do a clean fork and start over again doing 2 PRs.

Go for it. Just close this PR and open 2 new ones from 2 different branches. You don't need to fork the entire repo again. Just `git checkout master` to go back to where you branched and `git branch -D hidden-node-fix` to delete that branch forcibly (because you have unmerged changes, otherwise `gir branch -d ....`).

> seeding up drawing by detaching the columns from the dom and reattaching them after the drawing operation

May I make a suggestion? Create a branch for it, do a single one-line no-real-change commit, then create a `[WIP]` PR. Let's discuss that before you get too deeply into it. There are some side effects to consider.
